editor/plugins/theme_editor_plugin.cpp

```diff
@@ -1907,268 +1943,1291 @@ ThemeItemEditorDialog::ThemeItemEditorDialog() {
-	item->set_text(0, TTR("Has,Many,Options"));
-	item->set_range(0, 2);
-
-	main_hb->add_theme_constant_override("separation", 20 * EDSCALE);
-
 	theme_edit_dialog = memnew(ThemeItemEditorDialog);
 	theme_edit_dialog->hide();
-	add_child(theme_edit_dialog);
+	top_menu->add_child(theme_edit_dialog);
+
+	HSplitContainer *main_hs = memnew(HSplitContainer);
+	main_hs->set_v_size_flags(SIZE_EXPAND_FILL);
+	add_child(main_hs);
+
+	VBoxContainer *preview_tabs_vb = memnew(VBoxContainer);
+	preview_tabs_vb->set_h_size_flags(SIZE_EXPAND_FILL);
+	preview_tabs_vb->set_custom_minimum_size(Size2(520, 0) * EDSCALE);
+	preview_tabs_vb->add_theme_constant_override("separation", 2 * EDSCALE);
+	main_hs->add_child(preview_tabs_vb);
+	HBoxContainer *preview_tabbar_hb = memnew(HBoxContainer);
+	preview_tabs_vb->add_child(preview_tabbar_hb);
+	preview_tabs_content = memnew(PanelContainer);
+	preview_tabs_content->set_v_size_flags(SIZE_EXPAND_FILL);
+	preview_tabs_content->set_draw_behind_parent(true);
+	preview_tabs_vb->add_child(preview_tabs_content);
+
+	preview_tabs = memnew(Tabs);
+	preview_tabs->set_tab_align(Tabs::ALIGN_LEFT);
+	preview_tabs->set_h_size_flags(SIZE_EXPAND_FILL);
+	preview_tabbar_hb->add_child(preview_tabs);
+	preview_tabs->connect("tab_changed", callable_mp(this, &ThemeEditor::_change_preview_tab));
+	preview_tabs->connect("right_button_pressed", callable_mp(this, &ThemeEditor::_remove_preview_tab));
+
+	HBoxContainer *add_preview_button_hb = memnew(HBoxContainer);
+	preview_tabbar_hb->add_child(add_preview_button_hb);
+	add_preview_button = memnew(Button);
+	add_preview_button->set_text(TTR("Add Preview"));
+	add_preview_button_hb->add_child(add_preview_button);
+	add_preview_button->connect("pressed", callable_mp(this, &ThemeEditor::_add_preview_button_cbk));
+
+	DefaultThemeEditorPreview *default_preview_tab = memnew(DefaultThemeEditorPreview);
+	preview_tabs_content->add_child(default_preview_tab);
+	default_preview_tab->connect("control_picked", callable_mp(this, &ThemeEditor::_preview_control_picked));
+	preview_tabs->add_tab(TTR("Default Preview"));
+
+	preview_scene_dialog = memnew(EditorFileDialog);
+	preview_scene_dialog->set_file_mode(EditorFileDialog::FILE_MODE_OPEN_FILE);
+	preview_scene_dialog->set_title(TTR("Select UI Scene:"));
+	List<String> ext;
+	ResourceLoader::get_recognized_extensions_for_type("PackedScene", &ext);
+	for (List<String>::Element *E = ext.front(); E; E = E->next()) {
+		preview_scene_dialog->add_filter("*." + E->get() + "; Scene");
+	}
+	main_hs->add_child(preview_scene_dialog);
+	preview_scene_dialog->connect("file_selected", callable_mp(this, &ThemeEditor::_preview_scene_dialog_cbk));
+
+	theme_type_editor = memnew(ThemeTypeEditor);
+	main_hs->add_child(theme_type_editor);
+	theme_type_editor->set_custom_minimum_size(Size2(360, 0) * EDSCALE);
 }
 
 void ThemeEditorPlugin::edit(Object *p_node) {
-	if (Object::cast_to<Theme>(p_node)) {
+	if (p_node->is_class("Theme") && Object::cast_to<Theme>(p_node)) {
```

Comment:
What is the purpose of `p_node->is_class("Theme")`? You are basically checking type twice.

Reply:
Actually, a good question. What is the difference? Why we would use one over the other?

Reply:
`is_class()` uses strings and using strings is bad™. It's slower (not noticeably, but technically slower) and prone to mistakes. Not only you can misspell class name, when you rename class you might forget to rename all references and don't notice it until someone encounters some obscure bug. In case of `cast_to<>()`, any mistake results in compilation error. Also typing class name instead of string offers autocompletion.

Reply:
Makes sense. I guess it would be beneficial to clean up usages of `is_class` in other plugins that work like that.